src/components/getSelectedValue.ts

```diff
@@ -0,0 +1,24 @@
+import {get} from 'lodash'
+
+export const getSelectedValue = (
+  select: Record<string, string> | undefined,
+  document:
+    | {
+        [x: string]: unknown
+      }
+    | undefined
+): Record<string, unknown> => {
+  if (!select || !document) {
+    return {}
+  }
+
+  const selection: Record<string, string> = select || {}
+  const targetKeys = Object.keys(selection)
+  const selectedValue = targetKeys.reduce<Record<string, unknown>>((acc, key) => {
+    acc[key] = get(document, selection[key])
+
+    return acc
+  }, {})
```

Comment:
To answer your "Possible optimization" question, we could specifically look for empty references here and filter them out since it's unlikely that anyone wants/needs them to be returned by `options.select`:

```suggestion
  const selection: Record<string, string> = select || {}
  const selectedValue: Record<string, unknown> = {}
  for (const [key, path] of Object.entries(selection)) {
    let value = get(document, path)
    if(Array.isArray(value)) {
      // If there are references in the array, ensure they have `_ref` set, otherwise they are considered empty and can safely be ignored
      value = value.filter(
        item => 
          typeof item === 'object' 
          ? item?._type === 'reference' && '_ref' in item 
          : true
      )
    }
    selectedValue[key] = value
  }
```